Enable PVC as DataSource for PVC creation

j-griffith opened the pull request:
This enables the ability to specify and existing PVC as a DataSource in
a new PVC Spec (eg "clone" and existing volume).

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

> /kind feature

**What this PR does / why we need it**:

Currently the only valid input Kind for PVC DataSource is "VolumeSnapshot", this feature adds the ability to also specify a DataSource Kind of "PersistentVolumeClaim".

This results in the ability for a user to specify intent to clone and existing PVC, it does NOT implement cloning functionality in Kubernetes; that's left to the external csi provisioner and plugins.  It does however provide a mechanism to request a clone operation to the csi-provisioner and ultimately the csi plugin that the original PVC is allocated from.

Currently (as of 1.14) a user is able to enter a DataSource with any information they wish, however if it's not of the type "VolumeSnapshot" and the "VolumeSnapshotDataSource" feature gate is not enabled, the input is nilled and silently ignored.  This change follows the same pattern, via  the "VolumeDataSource" feature gate.

**Which issue(s) this PR fixes**:

Implements [enhancement/KEP 642](https://github.com/kubernetes/enhancements/pull/642)

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
This change enables a user to specify a DataSource/Kind of type "PersistentVolumeClaim" in their PVC spec.  This can then be detected by the external csi-provisioner and plugins if capable.
```

Commit:
Enable PVC as DataSource for PVC creation

This enables the ability to specify and existing PVC as a DataSource in
a new PVC Spec (eg "clone" and existing volume).

Commit:
Enable PVC as DataSource for PVC creation

This enables the ability to specify and existing PVC as a DataSource in
a new PVC Spec (eg "clone" and existing volume).

msau42 commented:
/assign @msau42 @jsafrane

msau42 commented on pkg/api/persistentvolumeclaim/util.go:
```diff
@@ -52,3 +52,17 @@ func volumeSnapshotDataSourceInUse(oldPVCSpec *core.PersistentVolumeClaimSpec) b
+
+func dataSourceIsEnabled(pvcSpec *core.PersistentVolumeClaimSpec) bool {
+	if pvcSpec.DataSource != nil {
+		if pvcSpec.DataSource.Kind == "PersistentVolumeClaim" && utilfeature.DefaultFeatureGate.Enabled(features.VolumeDataSource) {
```
These checks should also validate the Group

j-griffith replied:
Added to both PVC and Snapshot

msau42 commented on pkg/api/persistentvolumeclaim/util.go:
```diff
@@ -28,7 +28,7 @@ func DropDisabledFields(pvcSpec, oldPVCSpec *core.PersistentVolumeClaimSpec) {
 		pvcSpec.VolumeMode = nil
 	}
-	if !utilfeature.DefaultFeatureGate.Enabled(features.VolumeSnapshotDataSource) && !volumeSnapshotDataSourceInUse(oldPVCSpec) {
+	if !dataSourceIsEnabled(pvcSpec) && !dataSourceInUse(oldPVCSpec) {
```
can you add unit tests for this

j-griffith replied:
Done

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -416,11 +416,14 @@ type PersistentVolumeClaimSpec struct {
-	// a new volume and data will be restored to the volume at the same time.
-	// If the provisioner does not support VolumeSnapshot data source, volume will
+	// This field can be used to specify either:
+	// * An existing VolumeSnapshot
```
Be specific about the exact object types supported, ie "snapshot.storage.k8s.io/VolumeSnapshot"

j-griffith replied:
Sure, I'll reword this to explicitly call out the API

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -416,11 +416,14 @@ type PersistentVolumeClaimSpec struct {
+	// a new volume based on the contents of the specified PVC or Snapshot.
+	// If the provisioner does not support the specified data source, the volume will
 	// not be created and the failure will be reported as an event.
 	// In the future, we plan to support more data source types and the behavior
```
Add a note that both of these sources are alpha

msau42 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1524,6 +1524,7 @@ var supportedVolumeModes = sets.NewString(string(core.PersistentVolumeBlock), st
 
 var supportedDataSourceAPIGroupKinds = map[schema.GroupKind]bool{
 	{Group: "snapshot.storage.k8s.io", Kind: "VolumeSnapshot"}: true,
+	{Group: "", Kind: "PersistentVolumeClaim"}:                 true,
```
I think core objects should still have a group

j-griffith replied:
@msau42 It seemed redundant to require it based on the definition here:  https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L4062-L4065

The object checks default to the core type so it's not needed for any of the implementation.

xing-yang replied:
Core API's group is "".  That is why we had to make APIGroup optional.

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -416,11 +416,14 @@ type PersistentVolumeClaimSpec struct {
+	// must be enabled (VolumeSnapshotDataSource, VolumeDataSource)
+	// If the provisioner can support the specified data source, it will create
+	// a new volume based on the contents of the specified PVC or Snapshot.
+	// If the provisioner does not support the specified data source, the volume will
```
What happens if the provisioner is on an older version that doesn't understand the new field?

j-griffith replied:
If that's the case then the current CSI Provisioner includes a validation check of it's own and it fails provision for any DataSource Kind except Snapshot.  So the result would just be an error reported back from provisioning, the requested PVC should then go into error status.

xing-yang commented on pkg/features/kube_features.go:
```diff
@@ -434,6 +434,12 @@ const (
+	// alpha: v1.14
+	//
+	// Enable data source support (applies to all supported types, snapshots, populators and PVCs).
+	VolumeDataSource utilfeature.Feature = "VolumeDataSource"
```
This probably should be "VolumePVCDataSource".

j-griffith replied:
I'm open to changing to that; I was just following the pattern s/VolumeSnapshot/Volume/ but if it seems more clear to folks I can change the name.

As far as the comment, I need to update that; that's from the original version where I wanted to combine the feature gates into a single option (decided not to fight that battle) so I'll fix the comment.

Commit:
Update unit tests and feature name

Update the unit tests to include checks for incorrect APIGroup type in
PVC DataSource and change the name of the feature gate to be more clear:
s/VolumeDataSource/VolumePVCDataSource/

Commit:
Merge branch 'enable_pvc_datasource' of github.com:j-griffith/kubernetes into enable_pvc_datasource

j-griffith commented on pkg/features/kube_features.go:
```diff
@@ -436,10 +436,10 @@ const (
-	// Enable data source support (applies to all supported types, snapshots, populators and PVCs).
-	VolumeDataSource utilfeature.Feature = "VolumeDataSource"
+	// Enable support for specifying an existing PVC as a DataSource
+	VolumePVCDataSource utilfeature.Feature = "VolumePVCDataSource"
```
I think the old name was more clear; but this works for me if other prefer it.

j-griffith commented:
Sorry about the messiness; didn't squash the rebase commits and had around 420 commits added, tried to fix up by resetting to an older commit id and squashing from there.

j-griffith commented:
Closing this out and pointing to updated version without the rebase commits that added every label in the project:  https://github.com/kubernetes/kubernetes/pull/76913
